Update the models on object changes

antis81 opened the pull request:
# Description

This is the next big step to bring the models in sync' with changes in the Git repository. Those changes can origin from either inside MacGitver or a random application (like i.e. Git CLI).

When a Git repository changes, MacGitver must update its data models to work on the correct data.

For example a new commit, branch, tag, ... was created. MacGtiver has to react and update its data models, which in turn trigger an update on the bound views.

All changes are detected by a central instance - the repository manager `MacGitver::repoMan()` - which calls signals on a "detected change".
# Here is a list of possible events for orientation.
## Module "Repository"
- [x] repo opened
- [x] repo closed
- [x] repo activated
- [x] submodule activated
- [x] ~~HEAD of any "opened" repo changed~~
  - We don't need that!
## Module "History"
- [x] update inline references
  - [x] ref created
  - [x] ref deleted -> see `libMacGitverCore` [PR 22 there](https://github.com/macgitver/libMacGitverCore/pull/22)
  - [x] ~~ref renamed~~ -> This is not possible, because the reference name is the "lookup key". If it changes, it would generate a "refDeleted" and a "refCreated" event.
  - [x] ref moved (the "reference target" changed)
  - [x] HEAD changed (or any symbolic reference) -> :exclamation: See [here](https://redmine.macgitver.org/issues/237)
- [x] ~~update commit list~~ -> Needs to be realized in new PR.
## Module "RefsViews"
- [x] ref created
- [x] ref deleted
- ~~ref renamed~~ -> see  above in "Module "History""  
- [x] ref moved (detects also a changed HEAD)
## Status-View
- [x] ref moved (detects also a changed HEAD)
  - see `libMacGitverCore` [PR 22 there](https://github.com/macgitver/libMacGitverCore/pull/22)
## Module "Work-Tree / Index"
### WorkTree and Index/Stage View

Updating the work tree and the index has to go in another PR. Currently MGV does not react to changes in the working tree or index.
# Migration of the "Reference View"

Please read dicussion [in Redmine](https://redmine.macgitver.org/issues/213)

The reference view should be migrated to better integrate `RM::RepoMan`. Changing this now would make this PR unnecessarily complex.
# Other changes / improvements
## Reference View
- :+1: The reference tree is now expanded in initial state.
- :+1: The reference tree will no more collapse unexpectedly after changes.
## History View
- :+1: Inline References are now sorted alphanumerically by type (fixed priorities) - yeah!
  - If you're interested look here: https://github.com/macgitver/MacGitverModules/commit/7b9d2ced67cf197fbeb5205aafbf2e79ddbe00b1#diff-53684d014451b869081860fdde26e03bR89
- :+1: Fixed recognition of status changes in inline references. 
- :+1: The "default" diff behaviour is now "diff to all parents". By default, the diff view shows the changes in the selected commit.
  - Technically, there is no default behaviour. "All parents" is just the first entry in the list. Setting a default behaviour is a candidate for the history settings.
- :+1: Display a detached HEAD

antis81 committed:
typo

antis81 committed:
update history model, when a git reference was created

antis81 committed:
privatized appending history entries

antis81 committed:
moved assertion to where it belongs

antis81 committed:
simplified readability of HistoryModel::scanInlineRefs()

antis81 commented:
@scunz Would be nice, if you could verify this change bfd7051.

antis81 committed:
added some documentation

antis81 committed:
Ref-Model: added method to lookup a RefScope item for a Git::Reference

antis81 committed:
Ref-Model: added internal method to get a model index for a random RefItem

antis81 committed:
Ref-Model: added internal methods to insert a Git::Reference into the tree

antis81 committed:
Ref-Model: actually clear the reference tree in case of a model reset

antis81 committed:
Ref-Model: outsourced creation of references during model reset

antis81 committed:
Ref-Model: always create the RefScope items in a fixed order

antis81 committed:
Ref-Model: react to added references using RM::EventInterface

This is to be optimized. Currently it triggers a "re-lookup" of the added reference, using its full qualified name. Instead, the RM::Ref object should be used directly. This needs a bigger re-design of the current code model.

antis81 commented:
@scunz :+1: Some good news:
In the current implementation state, the `HistoryModel` and `ReferenceModel` both are aware of "reference creation". Thus, independent of the source (inside/outisde MGV), the views will get updated correctly based on the RepoMan events.

When there's light :bulb:, there's also some shadow:
Seems like the `RM::EventsInterface` needs extensions. Currently I don't get a signal when ...
1. ... HEAD was set (i.e. the active branch changed).
2. ... a reference was removed
3. ... an object was created (i.e. a new commit)

:question: Do we need definitions for them? Or is there something I overlooked here?

antis81 committed:
minor cleanup

antis81 committed:
RefsView: expand tree in initial state

antis81 committed:
Commit Dialog: Show errors during commit as warning message.

antis81 committed:
History-Diff: Moved "diff to parent commits" to top.

This sets "Diff To: All parents" as "default" behaviour. Technically, there is no default, but it is the first entry in the list. Selecting a default behaviour is a possible candidate for "History Settings".

antis81 committed:
RefsView: added RefsViewDelegate

antis81 committed:
RefsView: Set item delegate for tree view.

antis81 committed:
RefsView: return the base color for RowBgGradientRole

The gradient will be created in the item delegate.

antis81 committed:
RefsView: let the row background be drawn by the item delegate

antis81 committed:
RefItem: removed unused parameter

antis81 committed:
RefsView: first steps to migrate to RM::Ref

antis81 committed:
updated license info

antis81 committed:
fixup

antis81 committed:
History-Module: added "less than" comparator for "inline references"

Also added the mIsDetached flag, that will be used in upcoming commits.

antis81 committed:
fixed a deprecated warning

antis81 committed:
HistoryModel: outsourced internal update of "inline references"

antis81 committed:
HistoryModel: fixed comparison of "inline references" while refreshing

antis81 committed:
minor cleanup

antis81 committed:
HistoryModel: minor optimizations

antis81 committed:
HistoryModel: sort "inline references"

antis81 committed:
updated license info

antis81 committed:
HistoryModel: add an "inline reference" for a detached HEAD

antis81 committed:
HistoryModel: Use Qt5 QStringLiterals -> This breaks compatibility to Qt4

We don't need Qt4-Compatibility anyways and I don't want to keep that nasty stuff around.

antis81 committed:
HistoryModel: a stash is never "active"

antis81 committed:
HistoryModel: minor optimization

antis81 committed:
HistoryModel: react to symbolic reference changes

This is currently not triggered and needs work in RM::RepoMan.

antis81 committed:
HistoryView: mark a detached HEAD in a red color

antis81 commented:
I'm somewhat excited of what is available in this PR :+1:. It brings MGV a big step closer to a state, that's "publicly presentable". Well, some heavy weight bugs are still waiting to be fixed (especially memory consumption in the DiffViewer) ... but that's the next big story.

After completing some important things, I'm going to merge soon.

@scunz If you find any time, please help with the "refresh" topic. The most important parts happen in libMacGitverCore. I contribute my local work there as well - you shall see. At least, now we have real consumers for our events.

@BugSniffer If you want to know why I talk confusing stuff sometimes - just look here :smile:.

scunz commented on History/HistoryEntry.h:
```diff
@@ -61,9 +61,86 @@ struct HistoryInlineRef
+    }
+};
+
+typedef QList< HistoryInlineRef > HistoryInlineRefs;
```
Why is this a list now, instead of a vector?

antis81 replied:
Actually this is a result of "lots of changes" :) - Sane Reason: Because `std::sort()` runs over it. Is QVector good for sorting as well?

scunz commented on History/HistoryEntry.h:
```diff
@@ -61,9 +61,86 @@ struct HistoryInlineRef
+
+typedef QList< HistoryInlineRef > HistoryInlineRefs;
+
+struct HistoryInlineRef_LessThan
```
_If_ this is necessary at all, can it be done in private? (i.e. Move it to the .cpp)

antis81 replied:
Yes it can. Why if? The alternative would be to write an `operator <()`.

scunz replied:
Yes, an `operator<` would be more natural. But see may pending comment for another thing :)

antis81 replied:
Seems no commit is pending - I got ~10 :) - what do you mean?

scunz commented on History/HistoryEntry.h:
```diff
@@ -61,9 +61,86 @@ struct HistoryInlineRef
     bool        mIsStash    : 1;
+    bool        mIsDetached : 1;
+
+    bool operator ==(const HistoryInlineRef& other) const
```
What is the purpose of these operators? It looks like they're actually just implementing default behaviour.

antis81 replied:
Answer: Try removing this and write something like `if ( refA == refB ) ...`. The compiler fails in this case. Or am I missing something?

scunz replied:
Why should the compiler fail on that? It will (just like with member-wise copy and member-wise assignment) do a member-wise compare. It will only have problems for more complex operators like less-that, exclusive-or-with-self-assignment etc... :)

scunz commented on History/HistoryEntry.h:
```diff
@@ -61,9 +61,86 @@ struct HistoryInlineRef
+        }
+
+        return  mRefName    == other.mRefName       &&
+                mRefName    == other.mRefName       &&
```
Despite the fact that it compares `mRefName` twice. :)

antis81 replied:
Fixed :)

scunz commented on History/HistoryDiff.cpp:
```diff
@@ -52,6 +52,8 @@ HistoryDiff::HistoryDiff()
     mToolbar = new QToolBar;
 
     mDiffTo = new QComboBox;
+    mDiffTo->addItem( trUtf8( "All parents" ), DTT_AllParents );
```
This looks like a good workaround for the missing "make this setting persistent somewhere" bug :nail_care:

antis81 committed:
removed double comparison of mRefName

scunz commented on History/HistoryModel.cpp:
```diff
@@ -350,28 +356,52 @@ void HistoryModel::scanInlineReferences()
+            inlRef.mRefName = tr( "<recent stash>" );
             inlRef.mIsBranch = false;
-            inlRef.mIsCurrent = true;
+            inlRef.mIsCurrent = false;
```
That's wrong: _this_ stash is the most recent ones (You'll find the others be following the ref log).

antis81 replied:
Ah ... that actually confused me because of its ambiguous meaning. I'll change it back. It shouldn't have any effect though...

antis81 replied:
Fixed!

scunz commented on History/HistoryModel.cpp:
```diff
@@ -350,28 +356,52 @@ void HistoryModel::scanInlineReferences()
 
-        if (!refsById.contains(refs[ref])) {
-            refsById.insert(refs[ref], HistoryInlineRefs());
+        const Git::ObjectId &oid = refs[ref];
```
`refs` is a `Git::ResolvedRefs`, a `Git::ResolvedRefs` is a `QHash`.

Now, `refs` can either be `const` (which is not the case here), then `QHash`'s `const` operator `[]` never returns a reference but always a copy!
Or it can be non-const (which is the case here), then `operator[]` won't return a `const` reference...

What will the compiler do with that? I'd assume that it _might_ chose to create a temporary object on the stack and point the `oid` reference to it. It might further chose that the temporary object isn't required anymore after this line...

antis81 replied:
Yes sorry. Should have been aware of that!

scunz commented on History/HistoryModel.cpp:
```diff
@@ -380,53 +410,26 @@ void HistoryModel::scanInlineReferences()
 
-        if( !newRefs.count() )
+        // sort newRefs: no need to sort oldRefs as it is already sorted)
+        std::sort( newRefs.begin(), newRefs.end(), lt );
```
This sort thing is mega inefficient. Since std::sort cannot know anything about the used Qt container (which is the design, so it can be used by whatever container is compatible to the stl), it can only sort things in place.

However, there's two options:
1) We change the QList back to a QVector, then we'll basically have a C-style array of `HistoryInlineRef`, which is pretty memory effective (esp. since we don't modify that container once it is created). But sorting inside it means that the assignment operator of `HistoryInlineRef` has to be called 3 times for every swap. This can cause a lot of memory fragmentation. I'd rather avoid that for something that might be processed so frequently like this code.

2) We accept the change to QList. In that case, we're having one more indirection (QString + 64 bit most certainly doesn't fit into 64 bit). So, a `QList<HistoryInlineRef>` will essentially be a `QVector<HistoryInlineRef*>`. Adding to the first option, 2 or 3 `new` and `delete` calls for HistoryInlineRef, which has probably a size of 48 or 64 byte. Causing yet more memory fragmentation.

I'd assume that I was much more efficient to just sort the initial vector while creating it.

scunz commented on History/HistoryModel.h:
```diff
@@ -104,6 +114,15 @@ private slots:
+     * @internal
+     * @see RM::EventInterface
+     */
+    ///@{
```
This will probably uglify the doxygen output a lot - unless used _really_ consistently (and doesn't do what is expected anyway).

scunz commented on RefsViews/Branches/BranchesModel.cpp:
```diff
@@ -165,68 +171,110 @@ bool BranchesModel::hasChildren( const QModelIndex& parent ) const
+RefScope* BranchesModel::scopeForRef(const Git::Reference& ref) const
+{
+    RefItem* scope = NULL;
+    if ( ref.isLocal() )        scope = mRoot->children[0];
```
These magic indicies are not exactly very future proof. Better keep some local variables with their values in the class...

scunz commented on RefsViews/Branches/BranchesModel.hpp:
```diff
@@ -50,6 +57,23 @@ class BranchesModel : public QAbstractItemModel
     void gitError( const Git::Result& error );
 
+private slots:
+    /**
```
This documentation stuff should definitely go to the .cpp files.

scunz commented on RefsViews/Branches/BranchesModel.cpp:
```diff
@@ -165,68 +171,110 @@ bool BranchesModel::hasChildren( const QModelIndex& parent ) const
-            RefScope* scopeRemote = new RefScope( mRoot, tr( "Remote" ) );
-            RefScope* scopeOther = new RefScope( mRoot, tr( "Tags" ) );
-
             for( int i = 0; i < sl.count(); ++i )
```
This whole loop _should_ be removed (As soon as possible).

antis81 replied:
Actually I expect the `refCreated...` events to be fired here. For some reason they are not. If so, you are right. This would `double` the list.

scunz replied:
As a matter of fact, we'll have to deal with the situation that the creation of whatever type of model is populated does _not_ align with opening / reading the repository. I.e. the underlying ViewContext might have expired or the view is simply opened later. Or even just the _current_ state of the code.

The consequence of that is: We need to be able to feed (preferably a single view) with all events that lead to the current state. This makes this loop unnecessary. That's actually the reason, why this whole event thing isn't base solely on Qt signals/slots.

In fact, when thinking about this, it might turn out that a Qt signal / slot wrapper was actually a total-bullshit idea. Hooking into different levels of the `RepoManager` and requesting a full-update on creating that hook would probably be much saner.

antis81 replied:
Actually my comment was total bullshit :). The `refCreated` events are not fired on opening/reading the repo, because `repoEventsBlocked` prevents that for good reason.

Well ... what do you mean by "hook into levels of RepoManager". Isn't this the same thing in green?

scunz replied:
Yes, opening by itself shouldn't fire any events, actually. Though I'm not sure if that decision was right. However....

What we need is a way to force the actual model ("RepoMan") to send us a bunch of events that allow us to populate a given view from scratch. Once we've made that View-Concept of libBlueSky work, _it_ may decide to throw away all of our model (That we use to display) and asks us to recreate it when it is required.

What I was saying in my previous post, is that the Qt Signal/Slot mechanism doesn't allow us to do that. At least not for individual views. Let's say that we have 10 repositories open. Repository _A_ is currently displayed. Now, libBlueSky decides that repository _B_ has reached the maximum time allowed to be cached and just throws it away. The user now switches to either repository _B_ or to repository _C_ (which has never been open before). In both cases, libBlueSky will restore a ViewContext that doesn't have valid data assigned to it.

Now, have a look at the RepoMan, which is a huge tree of all open repositories and their individual objects (like References, Tags etc). What we can do with the Signal/Slot system is to _reset_ everything and have RepoMan send out events for everything. If every node in the RepoMan tree had the ability to have one or more observer(s) assigned to it (While the parent is implicitly an observer and event-proxy), we could hook up a ViewContext to any node inside the RepoMan hierarchy.

antis81 replied:
Basically I agree. I'm just practicing the "Art of small steps" here :) ... and I'm close to the finishing line - so I won't turn around at this point.

IOW:
- Step 1: Merge this PR, even if incomplete.
- Step 2: Currently the "RefModel" has nothing to do with RepoMan at all -> let's make them friends and Migrate "RefItem" to be actually aware of `RM::Ref` objects! This might become the observer you're talking about. I have local changes lying around here These became almost complete today morning. I'll open a new "PR" based on this merged state. Why? Because I don't want to destroy everything what's available here.
- Step 3: Make "RepoMan" the "Master-Model". This might be done in step 2, but seriously: let's discuss that [here](https://redmine.macgitver.org/issues/213). 

Btw.: I think the following "live-scenario" shows what you're describing about the "ViewContext". 

Just open a repository in MGV and close the "Repo-View" via the yellow `X` button. Then wait a few seconds and see what happens: The "Refs-View" will be suddenly cleared while the "HistoryView" stays populated. Didn't play around further, but it's saying enough ...

antis81 committed:
moved doxygen comments to cpp

antis81 committed:
HistoryModel: fixed temporary copy of object id

antis81 committed:
HistoryModel: changed back stash status to "isCurrent", as it contains the most recent stash

antis81 committed:
HistoryModel: react to refMoved signal

Most times this signal is send from RM when the HEAD reference changes.

antis81 committed:
HistoryModel: replaced less than comparator with operator <()

antis81 committed:
replaced QLatin1Literal with QStringLiteral

scunz commented on History/HistoryEntry.h:
```diff
@@ -61,9 +61,82 @@ struct HistoryInlineRef
+        // - tag
+        // - detached HEAD
+        // - current branch
+        // - local branch
```
Thinking about this a bit more, I'd like to insert "Remote Tracking Branch" into the sort order between current branch and local branches - Would that make sense?

antis81 replied:
Btw.: Isn't it better to make the operator static? Or is it static by definition?

antis81 replied:
> Thinking about this a bit more, I'd like to insert "Remote Tracking Branch" into the sort order between current branch and local branches - Would that make sense?

Yeah, that makes sense. Though it would look a bit weird. Actually this is worth another feature request to "get it right". See [here](https://redmine.macgitver.org/issues/238).

Btw.: Isn't it better to make the operator static? Or is it static by definition?

antis81 committed:
removed obsolete slot "onRefLinkChanged", which is handled by "onRefMoved"

antis81 committed:
Reference-Model: react to refMoved event

Currently forces a refresh of all the model data.

antis81 committed:
Reference-View: removed workaround to update the view manually

antis81 committed:
RepoManLogger: removed obsolete event "refLinkChanged"

antis81 committed:
History-Model: moved docs to source

antis81 committed:
History-Model: refresh inline references on a removed reference

This also is valid for a renamed reference!

antis81 committed:
Refs-Model: removed unused code

antis81 committed:
RefItem: added temporary helper methods

antis81 committed:
RefItem: removed obsolete method "setData"

antis81 committed:
Refs-Model: added temporary helper method to find invalid references

antis81 committed:
Refs-Model: react to deleted references

This is also valid for a renamed reference!

Minor downside: When a scope/namespace has no more children, it will still stay visible.

antis81 committed:
Refs-View: We don't need to manually update the RefsView anymore - finally!

antis81 commented:
@scunz That's the way aha, aha - I like it!! :smile:

Event detection for references is mostly complete. Basically nothing is missing. Just some possible optimizations:
- The "source repo" of an event is not checked due to the fact, we don't have that information available yet.
- If we check for the "HEAD" reference in `RM::Ref::selfRefresh`, we can add a `headMoved` event. This has the advantage, that the status-view will get less events.
- On quit, the event `refAboutToBeDeleted` is sent for every reference. Could be blocked here!
- Of course the 15 secs. refresh interval is annoying. But hey, it works!!!!! We need some mechanism though to trigger an "immediate update".

:boat: I'd like to merge soon. As always not everything is implemented.

Other bigger tasks worth a new PR:
- Adding commits to `HistoryView` on creation
- Migrate / Refresh the `WorkingTreeView`
- Refresh the `IndexView`

antis81 committed:
minor: removed debug output

antis81 committed:
Refs-Model: no need to check "ref" and "invalidRefs"

scunz commented on RefsViews/RefItem.hpp:
```diff
@@ -83,10 +91,13 @@ class RefNameSpace : public RefScope
 {
 public:
-    RefBranch( RefItem* p, const QString& t, const Git::Reference &ref );
+    explicit RefBranch(RefItem* p, const Git::Reference &ref);
```
Why is this `explicit`?

scunz commented on RefsViews/Branches/BranchesModel.hpp:
```diff
@@ -50,9 +56,25 @@ class BranchesModel : public QAbstractItemModel
+    QModelIndex index(RefItem* item) const;
+
+    void insertRef(bool notify, const Git::Reference& ref);
+    inline RefItem* insertNamespace(const bool notify, RefItem* parent, const QString& name);
```
If you declare something `inline`, then actually _make_ it inline by putting it into the header. These methods simply are not inline.

antis81 replied:
Thanks a lot for the hint! These lines are subject to change anyways as I really need to  do the migration now.

antis81 committed:
Refs-Model: inlined inline methods

antis81 committed:
RepoMan-Events: check, if the event-repo equals the model's repo

antis81 committed:
Refs-Model: made "scopeForRef" future proof

antis81 committed:
minor: updated copyright info

antis81 commented:
:+1: Migration of "RefItem" works 90% locally! Some topics to be solved, but nothing too bad ... you'll see.

The new feature branch `ngf/migrate-refitem` is based on this one - which is unusual, but a sane solution for this situation.

This PR must be merged together with PR 22 in `libMacGitverCore` -> 
[look here](https://github.com/macgitver/libMacGitverCore/pull/22)

@scunz I'd like to merge this one. Could you do some last review?

@BugSniffer FYI

antis81 commented:
@scunz Think we can close this one in favour to #99.
